cmd/clusterctl/pkg/client/cluster/move.go

```diff
@@ -0,0 +1,209 @@
+	}
+
+	o.uidToNode[ownerNode.identity.UID] = ownerNode
+	return ownerNode
+}
+
+// objToNode creates a node for the Kubernetes object received in input.
+// If the node corresponding to the Kubernetes object already exists as a virtual node detected when processing OwnerReferences,
+// the node is marked as Observed.
+func (o *objectGraph) objToNode(obj *unstructured.Unstructured) *node {
+	existingNode, found := o.uidToNode[obj.GetUID()]
+	if found {
+		existingNode.markObserved()
+		return existingNode
+	}
+
+	newNode := &node{
+		identity: objectReference{
+			APIVersion: obj.GetAPIVersion(),
+			Kind:       obj.GetKind(),
+			UID:        obj.GetUID(),
+			Name:       obj.GetName(),
+			Namespace:  obj.GetNamespace(),
+		},
+		dependents: make(map[*node]struct{}),
+		virtual:    false,
+	}
+
+	o.uidToNode[newNode.identity.UID] = newNode
+	return newNode
+}
+
+// getDiscoveryTypes returns the list of TypeMeta to be considered for the the move discovery phase.
+// This list includes all the types defines by the CRDs installed by clusterctl and the ConfigMap/Secret core types.
+func (o *objectGraph) getDiscoveryTypes() ([]metav1.TypeMeta, error) {
+	discoveredTypes := []metav1.TypeMeta{}
+
+	c, err := o.proxy.NewClient()
+	if err != nil {
+		return nil, err
+	}
+
+	crdList := &apiextensionsv1.CustomResourceDefinitionList{}
+	if err := c.List(ctx, crdList, client.MatchingLabels{clusterctlv1.ClusterctlLabelName: ""}); err != nil {
+		return nil, errors.Wrap(err, "failed to get the list of CRDs required for the move discovery phase")
+	}
+
+	for _, crd := range crdList.Items {
+		for _, version := range crd.Spec.Versions {
+			if !version.Storage {
+				continue
+			}
+
+			discoveredTypes = append(discoveredTypes, metav1.TypeMeta{
+				Kind: crd.Spec.Names.Kind,
+				APIVersion: metav1.GroupVersion{
+					Group:   crd.Spec.Group,
+					Version: version.Name,
+				}.String(),
+			})
+		}
+	}
+
+	discoveredTypes = append(discoveredTypes, metav1.TypeMeta{Kind: "Secret", APIVersion: "v1"})
+	discoveredTypes = append(discoveredTypes, metav1.TypeMeta{Kind: "ConfigMap", APIVersion: "v1"})
+
+	return discoveredTypes, nil
+}
+
+// Discovery reads all the Kubernetes objects existing in a namespace for the types received in input, and then adds
+// everything to the objects graph.
+func (o *objectGraph) Discovery(namespace string, types []metav1.TypeMeta) error {
+	c, err := o.proxy.NewClient()
+	if err != nil {
+		return err
+	}
+
+	selectors := []client.ListOption{
+		client.InNamespace(namespace),
+	}
+
+	for _, typeMeta := range types {
+		objList := new(unstructured.UnstructuredList)
+		objList.SetAPIVersion(typeMeta.APIVersion)
+		objList.SetKind(typeMeta.Kind)
+
+		if err := c.List(ctx, objList, selectors...); err != nil {
+			if apierrors.IsNotFound(err) {
+				continue
+			}
+			return errors.Wrapf(err, "failed to list %q resources", objList.GroupVersionKind())
+		}
+
+		for _, obj := range objList.Items {
+			o.addObj(obj.DeepCopy())
```

Comment:
(non-blocking) why the deep copy?

Reply:
I'm guessing because it returns a pointer

Reply:
If I use `&obj` I get 
```
cmd/clusterctl/pkg/client/cluster/move.go:206:14: Using a reference for the variable on range scope `obj` (scopelint)
```
So I used `obj.DeepCopy()`

Reply:
Oh, I'd do this:
```go
for i := range objList.Items {
  obj := objList.Items[i]
  o.addObj(&obj)
}
```